Java 16 Records breaking build-frontend

I am using Java 16  Records and there seems to be an issue doing a build with production mode enabled. 

POM Versions...
 <java.version>16</java.version>
 <vaadin.version>19.0.6</vaadin.version>

Run this on command line...
mvn -Pproduction package

When the build-frontend runs...
com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend

I get this error...
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend (default) on project artemis: Execution default of goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend failed: Record requires ASM8 -> [Help 1]

Looks like the bytebuddy code generator is not working with Java 16 records?

Hi @miguelatvaadin 

** jdk version...

openjdk version "16" 2021-03-16
OpenJDK Runtime Environment Corretto-16.0.0.36.1 (build 16+36)
OpenJDK 64-Bit Server VM Corretto-16.0.0.36.1 (build 16+36, mixed mode, sharing)



** Try adding this Record to your code to reproduce...

@JsonIgnoreProperties(value = "password")
public record User(Integer accountId,
                   String email,
                   String firstName,
                   String lastName,
                   Role role,
                   UserStatus accountStatus,
                   LocalDate createdOn) {
}


** This is the output from maven...

[INFO] --- vaadin-maven-plugin:19.0.6:build-frontend (default) @ artemis ---
[INFO] Scanning classes to find frontend configurations and dependencies...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14.400 s
[INFO] Finished at: 2021-05-06T19:28:19+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend (default) on project artemis: Execution default of goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend failed: Record requires ASM8 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: **Failed to execute goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend (default) on project artemis: Execution default of goal com.vaadin:vaadin-maven-plugin:19.0.6:build-frontend failed: Record requires ASM8**
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)


** Now if I swap out the Record above with a regular java class, vaadin-maven-plugin:19.0.6:build-frontend works without any  issues.

thanks!

So it appears vaadin FrontendDependencies class calls bytebuddy. And when byte buddy scans an instance of a Record it fails. Looks like the Opcodes version is set to Opcodes.ASM7 and not ASM8. Not sure how this is set internally but looks like its the source of the problem.

Caused by: java.lang.UnsupportedOperationException: Record requires ASM8
    at net.bytebuddy.jar.asm.**ClassVisitor**.visitRecordComponent (ClassVisitor.java:305)
    at net.bytebuddy.jar.asm.ClassReader.readRecordComponent (ClassReader.java:930)
    at net.bytebuddy.jar.asm.ClassReader.accept (ClassReader.java:708)
    at net.bytebuddy.jar.asm.ClassReader.accept (ClassReader.java:401)
    at com.vaadin.flow.server.frontend.scanner.**FrontendDependencies**.visitClass (FrontendDependencies.java:556)
    at com.vaadin.flow.server.frontend.scanner.FrontendDependencies.visitClass (FrontendDependencies.java:568)


![image](https://user-images.githubusercontent.com/78946/117415476-43fe2a00-af10-11eb-9dc6-4c28834d75c3.png)

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendAnnotatedClassVisitor.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendAnnotatedClassVisitor.java
@@ -60,7 +60,7 @@ final class FrontendAnnotatedClassVisitor extends ClassVisitor {
      *            The annotation class name to visit
      */
     FrontendAnnotatedClassVisitor(ClassFinder finder, String annotationName) {
-        super(Opcodes.ASM7);
+        super(Opcodes.ASM8);
         this.finder = finder;
         this.annotationName = annotationName;
         if (!annotationDefaults.containsKey(annotationName)) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendClassVisitor.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendClassVisitor.java
@@ -67,7 +67,7 @@ final class FrontendClassVisitor extends ClassVisitor {
 
     private final class FrontendMethodVisitor extends MethodVisitor {
         public FrontendMethodVisitor() {
-            super(Opcodes.ASM7);
+            super(Opcodes.ASM8);
         }
 
         // We are interested in the new instances created inside the method
@@ -143,7 +143,7 @@ final class FrontendClassVisitor extends ClassVisitor {
      */
     FrontendClassVisitor(String className, EndPointData endPoint,
             boolean themeScope) { // NOSONAR
-        super(Opcodes.ASM7);
+        super(Opcodes.ASM8);
         this.className = className;
         this.endPoint = endPoint;
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/RepeatedAnnotationVisitor.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/RepeatedAnnotationVisitor.java
@@ -27,7 +27,7 @@ import net.bytebuddy.jar.asm.Opcodes;
  */
 abstract class RepeatedAnnotationVisitor extends AnnotationVisitor {
     RepeatedAnnotationVisitor() {
-        super(Opcodes.ASM7);
+        super(Opcodes.ASM8);
     }
 
     @Override

